### PR TITLE
Fix dataflow predicate analysis for pattern operations

### LINKED ISSUE #2185: Handle DiscardPattern operation kind in flow-analysis package
IDiscardPatternOperation, OperationKind.DiscardPattern and IDeclarationPatternOperation.MatchedType were added in Microsoft.CodeAnalysis 3.0. We had to disable few asserts in flow analysis package to allow for these in features/dataflow branch targeting 2.9.0 bits. This issue tracks re-enabling those asserts, building Roslyn.sln to come up with minimal repro for these asserts and fix and add unit tests.

### LINKED ISSUE #3716: CA1062 False positive
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v3.0.0 (Latest)

#### Diagnostic ID

[CA1062](https://docs.microsoft.com/visualstudio/code-quality/ca1062)

#### Repro steps

I've a small abstract class that represents a chess piece, with the following Equals implementation.

image

#### Expected behavior

If I'm not missing something, there shouldn't be a warning.

1. Assume the passed `obj` is null.
1. `obj is Piece p` should be false.
1. `!(obj is Piece p)` should be true.
1. The right of the `||` operator won't be evaluated.

#### Actual behavior

A warning is issued.

### PR DESCRIPTION
**Pull request by mavasani**
Fixes #3716

Also handle the new discard and recursive patterns correctly.
Fixes #2185

### CODE REVIEW

File: src/Utilities/Compiler/Extensions/IOperationExtensions.cs [hunk 1]
```diff
@@ -512,6 +512,10 @@ public static bool IsWithinLambdaOrLocalFunction(this IOperation operation)
         {
             return pattern switch
             {
+#if CODEANALYSIS_V3_OR_BETTER
```
Comment (dotpaul) [inline src/Utilities/Compiler/Extensions/IOperationExtensions.cs:515]: Did you mean to target 2.9.x?

Comment (mavasani) [reply src/Utilities/Compiler/Extensions/IOperationExtensions.cs:515]: Ah yes, thanks!

Comment (mavasani) [reply src/Utilities/Compiler/Extensions/IOperationExtensions.cs:515]: Actually, no. These new IOperation APIs have been introduced in Microsoft.CodeAnalysis v3.3

Comment (dotpaul) [reply src/Utilities/Compiler/Extensions/IOperationExtensions.cs:515]: Maybe it should be named CODEANALYSIS_V3_3_OR_BETTER? Dunno where that's defined.

Comment (mavasani) [reply src/Utilities/Compiler/Extensions/IOperationExtensions.cs:515]: Sure, I can clean that up in a separate PR once this goes in. That works?

Comment (mavasani) [reply src/Utilities/Compiler/Extensions/IOperationExtensions.cs:515]: @dotpaul Any further feedback?

**Comment by mavasani** [comment]
Thanks @dotpaul

**Pull request merged**